### PR TITLE
[bsp/nuvoton] Support NuMaker-M2354 Cortex-M23 platform.

### PR DESCRIPTION
**Pull request by wosayttn**
## 拉取/合并请求描述：(PR description)

[

1. Support NuMaker-M2354 BSP.
2. Remove --c99 in components\drivers\spi\SConscript due to keil armclang doesn't support the option.

   Known Issue:
   Currently, ESP8266 Wi-Fi at-device using UART-RX-DMA mode cant workable due to backward-compatible issue. See #3293 #4305

]

以下的内容不应该在提交PR时的message修改，修改下述message，PR会被直接关闭。请在提交PR后，浏览器查看PR并对以下检查项逐项check，没问题后逐条在页面上打钩。
The following content must not be changed in the submitted PR message. Otherwise, the PR will be closed immediately. After submitted PR, please use a web browser to visit PR, and check items one by one, and ticked them if no problem.

### 当前拉取/合并请求的状态 Intent for your PR

必须选择一项 Choose one (Mandatory):

- [ ] 本拉取/合并请求是一个草稿版本 This PR is for a code-review and is intended to get feedback
- [x] 本拉取/合并请求是一个成熟版本 This PR is mature, and ready to be integrated into the repo

### 代码质量 Code Quality：

我在这个拉取/合并请求中已经考虑了 As part of this pull request, I've considered the following:

- [x] 已经仔细查看过代码改动的对比 Already check the difference between PR and old code
- [x] 代码风格正确，包括缩进空格，命名及其他风格 Style guide is adhered to, including spacing, naming and other styles
- [x] 没有垃圾代码，代码尽量精简，不包含`#if 0`代码，不包含已经被注释了的代码 All redundant code is removed and cleaned up
- [x] 所有变更均有原因及合理的，并且不会影响到其他软件组件代码或BSP All modifications are justified and not affect other components or BSP
- [x] 对难懂代码均提供对应的注释 I've commented appropriately where code is tricky
- [x] 本拉取/合并请求代码是高质量的 Code in this PR is of high quality

### CODE REVIEW

**Comment by Guozhanxin** [comment]
@wosayttn

Hi Wayne，

I’m Guo Zhanxin from Shanghai RuiSide Electronic Technology Co., Ltd. Today we plan to organize an online meeting to discuss "RTT serial port framework related issues". Bernard.Xiong said that you have mentioned some serial port related issues before, let me invite you.
Do you have time at 19:00 this evening?

Topic: https://docs.qq.com/doc/DQlFHbkFhb3d2Y1Rp

**Comment by wosayttn** [comment]
@Guozhanxin 

Thanks for your inviting, how to join the on-line meeting?

For #4305, I think we should revert previous version temporarily, because new modification effects basic serial-receiving using UART-RX-DMA. The root cause is rx_notice counting semaphore's value is not enough for getting a byte if UART-RX-DMA is enabled. The new routine always waits stored bytes in FIFO so that it hold up the at_client_getchar procedure.

**Comment by Guozhanxin** [comment]
> @Guozhanxin
> 
> Thanks for your inviting, how to join the on-line meeting?
> 
> For #4305, I think we should revert previous version temporarily, because new modification effects basic serial-receiving using UART-RX-DMA. The root cause is rx_notice counting semaphore's value is not enough for getting a byte if UART-RX-DMA is enabled. The new routine always waits stored bytes in FIFO so that it hold up the at_client_getchar procedure.

@wosayttn Do you have Wechat or 腾讯会议？Or can you say a recommended software？

**Comment by wosayttn** [comment]
> > @Guozhanxin
> > Thanks for your inviting, how to join the on-line meeting?
> > For #4305, I think we should revert previous version temporarily, because new modification effects basic serial-receiving using UART-RX-DMA. The root cause is rx_notice counting semaphore's value is not enough for getting a byte if UART-RX-DMA is enabled. The new routine always waits stored bytes in FIFO so that it hold up the at_client_getchar procedure.
> 
> @wosayttn Do you have Wechat or 腾讯会议？Or can you say a recommended software？

WeChat, My WeChat ID is the same with github username.

**Comment by wosayttn** [comment]
Hi, @BernardXiong

      Could you give us comment for this pending PR? We need do some modification for merging?
      Currently, new following PRs are blocking.

File: components/drivers/spi/SConscript [hunk 1]
```diff
@@ -28,7 +28,7 @@ if GetDepend('RT_USING_SFUD'):
     if rtconfig.CROSS_TOOL == 'gcc':
         LOCAL_CCFLAGS += ' -std=c99'
     elif rtconfig.CROSS_TOOL == 'keil':
-        LOCAL_CCFLAGS += ' --c99'
```
Comment (BernardXiong) [inline components/drivers/spi/SConscript:31]: Please keep this flag or could you please provide a better way for Keil MDK with ARMCC?

Comment (wosayttn) [reply components/drivers/spi/SConscript:31]: The CM23 ARCH need to use mdk-armclang to build project. Current RT building script just only supports ARMCC, but armclang not.  I have no idea on rtconfig.CROSS_TOOL. I think I should rollback this modification and notice 'remove --c99' in board's README.md.

**Pull request merged**